### PR TITLE
Moe bf16 ep

### PR DESCRIPTION
**Pull request by grimoire**
backends/moe.py and nn/moe.py has been refactored.
Reuse token dispatcher in DLBlas

### CODE REVIEW

File: lmdeploy/pytorch/envs.py [hunk 1]
```diff
@@ -126,6 +126,9 @@ def _patched_get_env(
     # we don't need to read this, it would be passed to ray workers
     # If Ray is launched from outside, it may fail to access the environment variables.
     os.getenv('DEEPEP_MAX_BATCH_SIZE', None)
+    os.getenv('DEEPEP_MAX_TOKENS_PER_RANK', None)
```
Comment (lvhan028) [inline lmdeploy/pytorch/envs.py:129]: Do we need to set those envs manually?

File: lmdeploy/pytorch/backends/cuda/moe/default.py [hunk 2]
```diff
@@ -0,0 +1,473 @@
...
+
+def build_deepep_moe(
+    low_latency_mode: bool,
+    ep_size: int,
+    ep_group: dist.ProcessGroup,
+    num_experts: int,
+    hidden_dim: int,
+    top_k: int,
+    layer_idx: int = 0,
+    chunk_size: Optional[int] = 32 * 1024,
```
Comment (lvhan028) [inline lmdeploy/pytorch/backends/cuda/moe/default.py]: `chunk_size` is not used by `FusedMoENormal`

File: lmdeploy/pytorch/envs.py (hunk 1 above)
Comment (grimoire) [reply lmdeploy/pytorch/envs.py:129]: DLBlas read these vars to build buffer.
https://github.com/DeepLink-org/DLBlas/blob/1710a860f654ddf50907251ec51670910368ee45/dlblas/layers/moe/token_dispatcher.py#L43

File: lmdeploy/pytorch/nn/moe/base.py [hunk 3]
```diff
@@ -0,0 +1,321 @@
...
+        raise NotImplementedError
+
+    def wait(self, state: Dict):
+        """wait."""
+        raise NotImplementedError
+
+    @property
+    def forward_dptp(self) -> MoEForwardDPTP:
+        """Forward dptp."""
+        raise NotImplementedError
```
Comment (lvhan028) [inline lmdeploy/pytorch/nn/moe/base.py]: Every derived class from FusedMoEBase calls `init_dist_args` to initialize `self._forward_dptp` and implements `def forward_dptp(self)` the same.
So, why not return `self._forward_dptp` here?

File: lmdeploy/pytorch/backends/cuda/moe/blocked_fp8.py [hunk 4]
```diff
@@ -0,0 +1,259 @@
...
+    # unpatch dlblas
+    dlblas.kernels.fused_moe_v3.use_deep_gemm = False
+
+    # unpatch deep_gemm
+    if func0_ is not None:
+        deep_gemm.get_col_major_tma_aligned_tensor = func0_
+        deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_masked = func1_
+    else:
+        del deep_gemm.get_col_major_tma_aligned_tensor
+        del deep_gemm.m_grouped_gemm_fp8_fp8_bf16_nt_masked
```
Comment (lvhan028) [inline lmdeploy/pytorch/backends/cuda/moe/blocked_fp8.py]: Is this for deepgemm compatiblity?

Comment (grimoire) [reply lmdeploy/pytorch/backends/cuda/moe/blocked_fp8.py]: Yes, this patch comes from main branch
https://github.com/InternLM/lmdeploy/blob/9096012a6cc48305b009d0f9d7f1c4a32c63512a/lmdeploy/pytorch/backends/cuda/moe.py#L416

Comment (lvhan028) [reply lmdeploy/pytorch/backends/cuda/moe/blocked_fp8.py]: Is this compatibility necessary? It adds maintenance overhead.
We can ask users to use the specified released version of deepgemm

File: lmdeploy/pytorch/backends/cuda/moe/default.py [hunk 5]
```diff
@@ -0,0 +1,469 @@
...
+            import deep_gemm  # noqa: F401
+        except ImportError:
+            logger.exception('DeepGEMM is required for DeepEP MoE implementation.')
+
+        # pre-allocate buffer
+        self.fusedmoe_build(True)
+
+    def update_weights(self, gate_up_weights: torch.Tensor, down_weights: torch.Tensor):
+        # gate_up_weights = gate_up_weights.transpose(1, 2).contiguous().transpose(1, 2)
+        # down_weights = down_weights.transpose(1, 2).contiguous().transpose(1, 2)
```
Comment (lvhan028) [inline lmdeploy/pytorch/backends/cuda/moe/default.py]: may remove the unused code

**Comment by grimoire** [comment]
Qwen3_30b_A3B_Thinking_2507  dp2 ep4

| dataset | version | metric | mode | 4144 |
|----- | ----- | ----- | ----- | -----|
| core_average | - | naive_average | gen | 66.25 |
|  | - | - | - | - |
| Instruction Following | - | - | - | - |
| IFEval | 353ae7 | Prompt-level-strict-accuracy | gen | 87.62 |
|  | - | - | - | - |
| General Reasoning | - | - | - | - |
| hle_llmjudge | 6ff468 | accuracy | gen | 10.80 |
| GPQA_diamond_repeat_4 | 772ea0 | accuracy (4 runs average) | gen | 70.83 |
|  | - | - | - | - |
| Math Calculation | - | - | - | - |
| aime2025_repeat_32 | 5e9f4f | accuracy (32 runs average) | gen | 85.62 |
|  | - | - | - | - |
| Knowledge | - | - | - | - |
| mmlu_pro | - | naive_average | gen | 79.93 |
|  | - | - | - | - |
| Code | - | - | - | - |
| lcb_code_generation_repeat_6 | b5b6c5 | pass@1 (6 runs average) | gen | 62.67 |

**Review by lvhan028 — APPROVED** [review]

**Comment by lvhan028** [comment]
https://github.com/InternLM/lmdeploy/pull/4084 has been merged. 
May merge the latest main and improve [lmdeploy/pytorch/backends/deepep_moe_checker.py](https://github.com/InternLM/lmdeploy/pull/4084/files#diff8b2804970ce6cbdbdbc80956f82e0a849c2b962a61da282b6ab601ee68ea4b85)  with the `singleton` decorator

File: lmdeploy/pytorch/kernels/cuda/fused_moe_ep.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+    expert_start_loc: torch.Tensor,
+    output_tensor: torch.Tensor,
+    m_indices: torch.Tensor,
+    output_index: torch.Tensor,
+):
+    BLOCK_E = 128  # token num of per expert is aligned to 128
+    num_warps = 8
+    num_experts = num_recv_tokens_per_expert.shape[0]
+    hidden_size = recv_x.shape[1]
+    # grid = (triton.cdiv(hidden_size, BLOCK_D), num_experts)
```
Comment (CUHKSZzxy) [inline lmdeploy/pytorch/kernels/cuda/fused_moe_ep.py]: Tiny nits, unused one?

File: lmdeploy/pytorch/kernels/cuda/fused_moe_ep.py [hunk 7]
```diff
@@ -0,0 +1,266 @@
...
+
+import torch
+import triton
+import triton.language as tl
+
+from .activation import silu_and_mul
+
+
+@triton.jit
+def _fwd_kernel_ep_scatter_1(
```
Comment (CUHKSZzxy) [inline lmdeploy/pytorch/kernels/cuda/fused_moe_ep.py]: Though from dlblas, should be their problem, `_fwd_kernel_ep_scatter_1`, `_fwd_kernel_ep_scatter_2` seems confusing function names.

File: lmdeploy/pytorch/nn/moe/base.py [hunk 8]
```diff
@@ -0,0 +1,321 @@
...
+from lmdeploy.pytorch.backends import OpType, get_backend
+from lmdeploy.pytorch.config import TPMode
+from lmdeploy.pytorch.distributed import get_dist_manager, get_tp_world_rank
+from lmdeploy.pytorch.model_inputs import get_step_ctx_manager
+
+
+class MoeType(Enum):
+    """Batch ecex type."""
+    Default = auto()
+    DSAsyncDecode = auto()
```
Comment (CUHKSZzxy) [inline lmdeploy/pytorch/nn/moe/base.py:19]: What's the meaning of DS?

**Review by CUHKSZzxy — COMMENTED** [review]
https://github.com/InternLM/lmdeploy/blob/main/lmdeploy/metrics/metrics_processor.py#L20

May consider use singleton to simply this as well, or leave it to me just as a reminder here.

File: lmdeploy/pytorch/nn/moe/base.py (hunk 8 above)
Comment (grimoire) [reply lmdeploy/pytorch/nn/moe/base.py:19]: Deepseek, I guess?

**Review by CUHKSZzxy — APPROVED** [review]
